app/src/main/proto/vss.proto

```diff
@@ -98,6 +98,64 @@ message PutObjectRequest {
 message PutObjectResponse {
 }
 
+message GetKeysSummaryRequest {
+
+  // store_id is a keyspace identifier.
+  // Ref: https://en.wikipedia.org/wiki/Keyspace_(distributed_data_store)
+  // All APIs operate within a single store_id.
+  // It is up to clients to use single or multiple stores for their use-case.
+  // This can be used for client-isolation/ rate-limiting / throttling on the server-side.
+  // Authorization and billing can also be performed at the store_id level.
+  string store_id = 1;
+
+  // A key_prefix is a string of characters at the beginning of the key. Prefixes can be used as
+  // a way to organize key-values in a similar way to directories.
+  //
+  // If key_prefix is specified, the response results will be limited to those keys that begin with
+  // the specified prefix.
+  //
+  // If no key_prefix is specified or it is empty, all the keys are eligible to be returned in the
+  // response.
+  optional string key_prefix = 2;
+
+  // last_evaluated_key is a pagination token.
+  //
+  // To query for the first page of GetKeysSummary, last_evaluated_key must not be specified.
+  //
+  // For subsequent pages, use the value that was returned as `last_evaluated_key` in the previous
+  // page's GetKeysSummaryResponse.
+  // It will be used as exclusive_start_key to perform the new query and only keys greater than
+  // the specified last_evaluated_key will be returned in the response.
+  optional string last_evaluated_key = 3;
+}
+
+message GetKeysSummaryResponse {
+
+  // Fetched versions along with the corresponding keys in the request.
+  repeated KeyValue key_versions = 1;
+
+  // last_evaluated_key is a pagination token, it is the key of the item where the operation stopped,
+  // inclusive of the current result set.
+  // Use this value to start a new GetKeysSummary operation, excluding this value in the new result
+  // by specifying this last_evaluated_key in the next request.
+  //
+  // If last_evaluated_key is empty, then the "last page" of results has been processed and there is
+  // no more data to be retrieved.
+  //
+  // If last_evaluated_key is not empty, it does not necessarily mean that there is more data in the
+  // result set. The only way to know when you have reached the end of the result set is when
+  // last_evaluated_key is empty.
+  //
+  // Caution: Clients must not assume a specific number of key_versions to be present in a page for
+  // paginated response.
+  optional string last_evaluated_key = 2;
+
+  // global_version is a sequence-number/version of the whole store.
+  //
+  // global_version is only returned in response for the first page of GetKeysSummaryResponse.
+  optional int64 global_version = 3;
```

Comment:
What if the global version is updated on the server before the call for the next page? Will the request fail?

Reply:
Currently, not really.

So the current correct usage of this api is:
1. read global_version from first response, store it in local_variable. 
2. apply all the key versions.
3. then apply the global_version from the previously stored local_variable. 

i should probably add this in doc on server-side as well, to help in usage.

Reply:
> Currently, not really.
> 
> So the current correct usage of this api is:
> 
> 1. read global_version from first response, store it in local_variable.
> 2. apply all the key versions. (load objects lazy or eager)

Could you describe the lazy case? Is it essentially marking the client-side version as out-of-date, requiring it to be fetched when it needs to read next?

> 3. then apply the global_version from the previously stored local_variable.
> 
> i should probably add this in doc on server-side as well, to help in usage.

Hmmm... I'm not sure of this addressed my concern around pagination. Couldn't the second page correspond to a different global version if another client updated the global version after fetching the first page but before fetching the second?

Reply:
just to clarify, since there's no way to do a batch-get of the entire store, there's no hope for getting a globally consistent snapshot incrementally, right?  so perhaps it doesn't matter that each page may have a different global version?

Reply:
>Is it essentially marking the client-side version as out-of-date, requiring it to be fetched when it needs to read next? 

Yes, this is optional but on client side we can store the new-version-number and old-version-number. On get of key, if there is a diff, we can make a service call and fetch latest instead of returning the local version.

>Couldn't the second page correspond to a different global version if another client updated the global version

The global version is only returned in first page of results. With that being said, there still exists a scenario where globalVersion/store is updated after first page's fetch. This constitutes a race-condition b/w ListKeyVersions on device-1 and put-operation on device-2.
In such a scenario, it is ensured that the api is used safely if client follows the above protocol.

I.e., the put requests and global-versioning still maintains correct versioning and does not let the primary store to get corrupted. (the api design decisions are made to enforce/encourage correct and safe usage by the client.)

It does not guarantee that client has fully updated view of store in all scenarios after refresh, it makes a best-effort to bring client to fully-updated state. In case of race-conditions, we make sure that versioning scheme is not compromised and subsequent refresh will bring client up-to-date. In case where client has slightly out-of-date view of key-versions, subsequent put-requests will fail and client will have to refresh. 

Current Api contract is as following: "Here are the minimum key-versions by `this` globalVersion. Which is globally consistent snapshot at instance of time when store's globalVersion was `this`"

I did not want to return global-version in every page, since client could easily end up misusing that api.

Reply:
just to be a bit more specific, here is my understanding:

- each page is internally consistent at a certain global version
- a later page can have a greater-than-or-equal global version to the previous page (i.e. monotonically increasing)

is that what you had in mind?

Reply:
> It does not guarantee that client has fully updated view of store in all scenarios after refresh, it makes a best-effort to bring client to fully-updated state. In case of race-conditions, we make sure that versioning scheme is not compromised and subsequent refresh will bring client up-to-date. In case where client has slightly out-of-date view of key-versions, subsequent put-requests will fail and client will have to refresh.

Couple things came to mind:

- Is this robust enough to handle the case where the put-operation on device-2 adds data for a new key that would have been returned to device-1 in the first page?
- Will the server essentially exclude that item from the second page?
- For lightning, does that leave us in a state for instance when device-2 may open a new channel but device-1 may not know it exists for making a payment until either a refresh or an incoming payment? Or for the latter would it fail because the channel is unknown (i.e., it won't bother querying the server)?

Reply:
@devrandom 
Not exactly, only first page contains the globalVersion. And the key-versions are guaranteed to be correct only for that global version. Since client would be using that global version in put requests, it will only work if there was no further update **during or after this api** request.

@jkczyz 
>Is this robust enough to handle the case where the put-operation on device-2 adds data for a new key that would have been returned to device-1 in the first page?

Yes, since client would be using an out-of-date globalVersion, a client-device not having updated view will not be able to write successfully. 
 
 > Will the server essentially exclude that item from the second page?

 Server does not do anything special, globalVersion ensures that client has complete and updated view. If some change has been made **during or after this api**, then client's globalVersion is rendered out-of-date.
   
> For lightning, does that leave us in a state for instance when device-2 may open a new channel but device-1 may not know it exists for making a payment until either a refresh or an incoming payment? Or for the latter would it fail because the channel is unknown (i.e., it won't bother querying the server)?

The key is, client is not able to perform any changes or updates if they don't have fully-updated view.
If the primary state has had any change, client-device needs to update/refresh to proceed with any operations.
This refresh could be triggered because of unrelated change or some periodic refresh on client-side.

Client-device could also ensure that they have latest state by just querying and comparing the globalVersion.
Depending on implementation, this could be triggered upfront i.e. before doing the actual work/operation.

>it won't bother querying the server

This cannot happen because in order to use anything or make any save client would need to have fully updated view.
This enforces client to have up-to-date view of LN state-machine.


Note that this multi-device usage is intended for end-user devices and not for concurrently running routing nodes of same LN-state machine. In these cases, we do expect some contention and overhead but our main goal is to ensure safe state machine transitions. And since these points of parallel use are user-facing, we expect them to be occasional but at the same time not very high frequency.  

Expect context switching to be from one device to another, and in worst case back to same device. But we should not expect back and forth again and again b/w same devices within a short interval. (Even though all transitions are ensured to be safe, there will be some overhead when switching and that should be fine)

Reply:
> Yes, since client would be using an out-of-date globalVersion, a client-device not having updated view will not be able to write successfully.
>
> Server does not do anything special, globalVersion ensures that client has complete and updated view. If some change has been made **during or after this api**, then client's globalVersion is rendered out-of-date.
> 

My question here was not around writing but rather reading subsequent pages. Say device-1 gets page-1 with:
- key-1
- key-2
- key-4

Then device-2 puts key-3. Would device-1 get key-3 on when retrieving page-2?

> The key is, client is not able to perform any changes or updates if they don't have fully-updated view. If the primary state has had any change, client-device needs to update/refresh to proceed with any operations. This refresh could be triggered because of unrelated change or some periodic refresh on client-side.
> 

But say the client just wants to form a payment path using their channels first hop candidates. No write is involved. So they would have an outdated view of the channels (i.e., missing the newly opened channel). I see that if they actually go through and try to make the payment using the path --which would involve a write -- it would fail because of the global version is out of date. But after fetching the data, would they just go through with making the payment? Or would they start over with path finding now that they are aware of the previously missing channel? This is more of a question for users of VSS rather than how it is implemented, I suppose.

Reply:
>Would device-1 get key-3 on when retrieving page-2?

It might or it might not, depending on DB implementation being used and key ordering. But it shouldn't matter according to the current process. (In both cases, a refresh is still required.)


> Second question

That is an implementation detail and yes it depends on how vss is being used and refreshed. 
One way to go about it is, client will have some buffer window for which it will assume that it has the latest view.
If client-state was refreshed within T-x time, assume it is latest and proceed.